### PR TITLE
Global ex

### PR DESCRIPTION
**Pull request by DenisBogusevich**

### CODE REVIEW

File: src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        body.put("status", HttpStatus.BAD_REQUEST);
+
+        List<String> errors = ex.getBindingResult().getAllErrors().stream()
+                .map(e -> getErrorMassage(e))
+                .toList();
+        body.put("errors", errors);
+        
+        return new ResponseEntity<>(body, headers, status);
+    }
+
```
Comment (pavlov-n) [inline src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java:37]: I think that you can to catch your EntityNotFoundException here

**Review by pavlov-n — APPROVED** [review]
Looks good. But i have one comment for you. Try to rebase your pr to other branches, because while you are not merged them, it's hard to understand what i need to check.
P.s.Who will check, say please, if i'm wrong

File: src/main/java/onlinebookstore/controler/BookControler.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Book controler", description = "Endpoints for managing books")
+@RequiredArgsConstructor
+@RestController
+@Validated
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/controler/BookControler.java]: Do you need this annotation?

File: src/main/java/onlinebookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package onlinebookstore.dto;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import java.math.BigDecimal;
+
+public record CreateBookRequestDto(
+        @NotNull
+        String title,
+        @NotNull
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/dto/CreateBookRequestDto.java]: better use NotBlank with strings

File: src/main/java/onlinebookstore/dto/CreateBookRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+        String author,
+        @NotNull
+        String isbn,
+        @NotNull
+        @Min(0)
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/dto/CreateBookRequestDto.java]: You can use PositiveOrZero

File: src/main/java/onlinebookstore/exception/DataProcessingException.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package onlinebookstore.exception;
+
+public class DataProcessingException extends RuntimeException {
+    public DataProcessingException(String s, Exception e) {
+        super(s, e);
+    }
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/exception/DataProcessingException.java]: Don’t  use abbreviations

File: src/main/java/onlinebookstore/repository/BookRepository.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package onlinebookstore.repository;
+
+import onlinebookstore.entity.Book;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface BookRepository extends JpaRepository<Book, Long> {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/repository/BookRepository.java]:
```suggestion
```

File: src/main/java/onlinebookstore/service/BookService.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    BookDto save(CreateBookRequestDto book);
+
+    List<BookDto> findAll(Pageable pageable);
+
+    BookDto findById(Long id);
+
+    void deleteById(Long id);
+
+    BookDto update(Long id, CreateBookRequestDto bookDto);
+
```
Comment (liliia-ponomarenko) [inline src/main/java/onlinebookstore/service/BookService.java]:
```suggestion
```

File: src/main/resources/db/changelog/changes/01-create-books-table.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                  type: VARCHAR(1000)
+              - column:
+                  name: cover_image
+                  type: VARCHAR(255)
+              - column:
+                  name: is_deleted
+                  type: boolean
+                  defaultValueBoolean: false
+                  constraints:
+                    nullable: false
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changes/01-create-books-table.yaml]: better add one empty line to the end of file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
good job! Let’s  fix a few mistakes;)

File: pom.xml [hunk 9]
```diff
@@ -48,14 +67,69 @@
...
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-maven-plugin</artifactId>
+            <version>4.23.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.hibernate</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>8.0.1.Final</version>
+        </dependency>
```
Comment (NazarSmith228) [inline pom.xml]: this is redundant, since you've added `spring-boot-starter-validation`

pls remove

File: pom.xml [hunk 10]
```diff
@@ -48,14 +67,69 @@
...
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>4.23.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-maven-plugin</artifactId>
+            <version>4.23.0</version>
+        </dependency>
```
Comment (NazarSmith228) [inline pom.xml]: this is a plugin - pls remove from dependencies section

File: pom.xml [hunk 11]
```diff
@@ -48,14 +67,69 @@
...
             <artifactId>lombok</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>4.23.0</version>
```
Comment (NazarSmith228) [inline pom.xml]: you don't need to specify the version here

File: pom.xml [hunk 12]
```diff
@@ -48,14 +67,69 @@
             <groupId>org.projectlombok</groupId>
             <artifactId>lombok</artifactId>
         </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
+            <artifactId>h2</artifactId>
+        </dependency>
```
Comment (NazarSmith228) [inline pom.xml:73]:
```suggestion
        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <scope>test</scope>
        </dependency>
```

File: src/main/java/onlinebookstore/controler/BookControler.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Operation(summary = "Update book by id")
+    @PutMapping("/{id}")
+    public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto bookDto) {
+        return bookService.update(id, bookDto);
+    }
+
+    @Operation(summary = "Create book")
+    @PostMapping
```
Comment (NazarSmith228) [inline src/main/java/onlinebookstore/controler/BookControler.java:48]: pls add CREATED response status

File: src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
+
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
```
Comment (NazarSmith228) [inline src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java:21]:
```suggestion
    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(
```

File: src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
Comment (NazarSmith228) [inline src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java:18]: `@ControllerAdvice` is missing

File: src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            HttpHeaders headers,
+            HttpStatus status,
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.BAD_REQUEST);
```
Comment (NazarSmith228) [inline src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: it makes no sense to add status to response body

pls remove

File: src/main/java/onlinebookstore/exception/DataProcessingException.java [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package onlinebookstore.exception;
+
+public class DataProcessingException extends RuntimeException {
+    public DataProcessingException(String s, Exception e) {
```
Comment (NazarSmith228) [inline src/main/java/onlinebookstore/exception/DataProcessingException.java]:
```suggestion
    public DataProcessingException(String message, Throwable cause) {
```

**Review by NazarSmith228 — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java (hunk 14 above)
Comment (DenisBogusevich) [reply src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java:21]: can`t add this anotation becouse of error

Comment (liliia-ponomarenko) [reply src/main/java/onlinebookstore/exception/CustomGlobalExceptionHandler.java:21]: Which error? It should  works! Let’s  discuss  in your chat

File: pom.xml (hunk 12 above)
Comment (liliia-ponomarenko) [reply pom.xml:73]: not fixed

**Review by liliia-ponomarenko — APPROVED** [review]
Good job! But let’s  fix a few mistakes  before merging ;)

**Pull request merged**